431203: CVE-2007-6698 openldap: slapd crash on NOOP control operation on entry in bdb storage

Source: upstream, http://www.openldap.org/its/index.cgi/Software%20Bugs?id=4925,
        http://www.openldap.org/its/index.cgi/Software%20Bugs?id=5358
http://www.openldap.org/devel/cvsweb.cgi/servers/slapd/back-bdb/modify.c.diff?r1=1.124.2.16&r2=1.124.2.17&f=h
http://www.openldap.org/devel/cvsweb.cgi/servers/slapd/back-bdb/modrdn.c.diff?r1=1.197&r2=1.198

Index: openldap-2.3.32/servers/slapd/back-bdb/modify.c
===================================================================
--- openldap-2.3.32.orig/servers/slapd/back-bdb/modify.c
+++ openldap-2.3.32/servers/slapd/back-bdb/modify.c
@@ -538,6 +538,8 @@ retry:	/* transaction retry */
 		} else {
 			rs->sr_err = LDAP_X_NO_OPERATION;
 			ltid = NULL;
+			/* Only free attrs if they were dup'd.  */
+			if ( dummy.e_attrs == e->e_attrs ) dummy.e_attrs = NULL;
 			goto return_results;
 		}
 	} else {
Index: openldap-2.3.32/servers/slapd/back-bdb/modrdn.c
===================================================================
--- openldap-2.3.32.orig/servers/slapd/back-bdb/modrdn.c
+++ openldap-2.3.32/servers/slapd/back-bdb/modrdn.c
@@ -729,6 +729,8 @@ retry:	/* transaction retry */
 		} else {
 			rs->sr_err = LDAP_X_NO_OPERATION;
 			ltid = NULL;
+			/* Only free attrs if they were dup'd.  */
+			if ( dummy.e_attrs == e->e_attrs ) dummy.e_attrs = NULL;
 			goto return_results;
 		}
 
